lib/src/model/notifications/local_notification_service.dart

```diff
@@ -0,0 +1,156 @@
+import 'package:lichess_mobile/src/model/notifications/info_notification.dart';
+import 'package:lichess_mobile/src/model/settings/general_preferences.dart';
+import 'package:lichess_mobile/src/utils/l10n.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'local_notification_service.g.dart';
+
+@Riverpod(keepAlive: true)
+LocalNotificationService localNotificationService(
+  LocalNotificationServiceRef ref,
+) {
+  return LocalNotificationService(ref, Logger('LocalNotificationService'));
+}
+
+class LocalNotificationService {
+  LocalNotificationService(this._ref, this._log);
+
+  final LocalNotificationServiceRef _ref;
+  final Logger _log;
+  final Map<int, void Function(String?, String?)> _callbacks = {};
+  final _notificationPlugin = FlutterLocalNotificationsPlugin();
+  final _receivePort = ReceivePort();
+
+  int currentId = 0;
+
+  Future<void> init() async {
+    _updateLocalisations();
+    _ref.listen(generalPreferencesProvider, (prev, now) {
+      if (prev!.locale == now.locale) return;
+      _updateLocalisations();
+    });
+
+    // hot reloading doesnt remove the port so we just make sure it doesnt exist before registering it
+    IsolateNameServer.removePortNameMapping('localNotificationServicePort');
+    IsolateNameServer.registerPortWithName(
+      _receivePort.sendPort,
+      'localNotificationServicePort',
+    );
+    _receivePort.listen(_onReceivePortMessage);
+
+    await _notificationPlugin.initialize(
+      InitializationSettings(
+        android: const AndroidInitializationSettings('logo_black'),
+        iOS: DarwinInitializationSettings(
+          requestAlertPermission: false,
+          requestBadgePermission: false,
+          requestSoundPermission: false,
+          notificationCategories: <DarwinNotificationCategory>[
+            ChallengeNotificationDetails.instance.darwinNotificationCategory,
+          ],
+        ),
+      ),
+      onDidReceiveNotificationResponse: _notificationResponse,
+      onDidReceiveBackgroundNotificationResponse:
+          _notificationBackgroundResponse,
+    );
+    _log.info('[Local Notifications] initialized');
+  }
+
+  void _updateLocalisations() {
+    final l10n = _ref.read(l10nProvider);
+    InfoNotificationDetails(l10n.strings);
+    ChallengeNotificationDetails(l10n.strings);
+  }
+
+  Future<int> show(LocalNotification notification) async {
+    final id = currentId++;
+    await _notificationPlugin.show(
+      id,
+      notification.title,
+      notification.body,
+      notification.notificationDetails,
+      payload: notification.payload,
+    );
+    _log.info(
+      '[Local Notifications] Sent notification: ($id | ${notification.title}) ${notification.body} (Payload: ${notification.payload})',
+    );
+
+    // if the notification implements LocalNotificaitonCallback
+    if (notification is LocalNotificationCallback) {
```

Comment:
Instead of checking an interface, you could simply add a nullable `callback` field to the `LocalNotification` class and check for its presence. That would avoid unnecessary casts.